src/eigen/preconditioners.jl

```diff
@@ -68,4 +70,11 @@ end
 
 function precondprep!(P::PreconditionerTPA, X)
     P.mean_kin = [real(dot(x, Diagonal(P.kin), x)) for x in eachcol(X)]
+    array_type = typeof(similar(X,eltype(X),size(P.mean_kin)))
+    P.mean_kin = convert(array_type, P.mean_kin)
 end
+
+#TODO: remove this if it implemented in GPUArrays
+import LinearAlgebra.dot
+using GPUArrays
+LinearAlgebra.dot(x::AbstractGPUArray, D::Diagonal,y::AbstractGPUArray) = x'*(D*y)
```

Comment:
🏴‍☠️! Put that in workarounds/ so we at least isolate places where we're bad

Reply:
So far this type of workaround happens twice (for `dot` and for the equality test `B!=I` in LOBPCG), but it should really go in CUDA or GPUArrays. I put it in workarounds/ for now but it should soon go away.